### ISSUE TITLE
Few questions about pretrained model.

### ISSUE THREAD

**Issue by OValery16** [alexandrosstergiou/Squeeze-and-Recursion-Temporal-Gates#8]
I am currently trying to reproduce some of your results.

Regarding your mtnet pretrained models,  in Table 1 of your paper, I see a softpooling 1 x 2 x 2. However in your code, it seems that you only use a maxpool (MTnet.py l 499). In the pretrained weights are you using a maxpool or a softpool for your results ?

![image](https://user-images.githubusercontent.com/17808941/109331270-f18ff580-785c-11eb-8d6c-e89937f90fab.png)

**Comment by alexandrosstergiou** [comment]
Hi @OValery16,

The results in the paper were indeed with SoftPool. However, I had used a patched PyTorch version because I was getting `NaN` values. That was because the code in this repo runs at mixed-precision (`float16`) while there is no half-precision implementation at the SoftPool CUDA version. So I did multiple castings to different types in my PyTorch version.

With that said, I also tested MaxPooling instead, and the results followed roughly the same trend as SoftPool. So this makes it easier for people to deploy the code.

On a different note, the MTNet paper is still in progress, so I am actively updating it to a newer/improved version.

Best,
Alex

**Comment by OValery16** [comment]
Hi @alexandrosstergiou,

Regarding your problem with the NaN value, I got similar issue for another layer. The solution I found is to use an apex directive:

```
@amp.float_function
def layer_forward_you_want_to_run_in_float(self,x):
      x=self.layer(x)
      return x
      
```

By doing that, you won't get any underflow issue. (the problem is that some of your softmax probabilities cannot hold on float 16).  I got the same issue in the past. I hope this solution will help you.

So, if I understood well, the pretrained weight provided in this repo come from a training where the main model uses maxpooling (not softpooling). Can you confirm me that ? (I want to test your pretained model,  so it is quite important for me that the network configuration fits the pretrained weight)

I have another question. I saw the file temporal_alignment.py which is not called anywhere. Can you tell me a bit more about it ? How does it work ? How do you handle the temporal alignment for a dataset such as HACS ? If I understand well your code, you use a small network to make sure that the input video is temporally aligned. How do you train such network ?

Best regards,

Olivier

**Comment by alexandrosstergiou** [comment]
Hi @OValery16 ,

Thanks for the snippet! I will see if I can integrate it and check if I'm still getting underflows from the CUDA SoftPool implementation.

Yes, basically if you call any of the MTNets without changing any parameters, the weights will load. Alternatively, the only possible problems would be in the `nn.MaxPool3d`[in line #409](https://github.com/alexandrosstergiou/Squeeze-and-Recursion-Temporal-Gates/blob/9d287a9ba8e5a245a308511d0523736f17eef8b1/network/MTnet.py#L409). The rest of the pooling operations are done with function calls, so there is no mismatch in the `state_dict`. 

The `temporal_alignment.py` file was just a proof of concept to test cyclic consistency. I have moved the code to the `soft_nnc` function in the `Squeeze_and_Recursion` class [lines](https://github.com/alexandrosstergiou/Squeeze-and-Recursion-Temporal-Gates/blob/9d287a9ba8e5a245a308511d0523736f17eef8b1/network/srtg_resnet.py#L269-L321). The point of the function is to test that the two inputs cycle-back to each other through their soft NN. In other words, every frame `i` in `x_1` should correspond to frame at position `i` in `x_2` and equally every frame in position `i` in `x_1` should correspond to frame `i` in `x_2`. 

Best,
Alex
